enhancements/api-review/add-ClusterImagePolicy-and-ImagePolicy-for-Signature-Verification.md

```diff
@@ -192,6 +196,46 @@ type FulcioCAWithRekor struct {
 	FulcioSubject PolicyFulcioSubject `json:"fulcioSubject,omitempty"`
 }
 
+// PKI defines the root of trust based on Root CA(s) and corresponding intermediate certificates.
+type PKI struct {
+	// CertificateAuthorityRootsData contains a certificate bundle PEM file containing one or more CA roots
+	// in the PEM format.
+	// caRootsData cannot be set if certificateChainData is set.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateChainData == ''", message="CertificateAuthorityRoots cannot be set if CertificateChain is set"
+	CertificateAuthorityRootsData []byte `json:"caRootsData,omitempty"`
+
+	// CertificateAuthorityIntermediatesData contains a certificate bundle PEM file containing one or more intermediate certificates
+	// in the PEM format.
+	// caIntermediatesData cannot be set if certificateChainData is set.
+	// caIntermediatesData requires CertificateAuthorityRoots to be set.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateChainData == ''", message="caIntermediatesData cannot be set if certificateChainData is set"
+	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateAuthorityRootsData != ''", message="caIntermediatesData requires caRootsData to be set"
+	CertificateAuthorityIntermediatesData []byte `json:"caIntermediatesData,omitempty"`
+
+	// CertificateIdentity must be set if any of the other fields are set.
+	// +optional
```

Comment:
Doesn’t this amount to the identity being mandatory?

---

But, also, see the comment on `CertificateIdentity`. What is the semantics we are proposing, _and then_ what are the k8s API evolution techniques we need to use?

Reply:
Maybe a rule like,

// +kubebuilder:validation:XValidation:rule="self.CertificateAuthorityRootsData == '' || self.PKICertificateSubject != nil", message="pkiCertificateSubject must be set if caRootsData is specified"

Reply:
`caRootsData` is now required, so I think that implies to `PKICertificateSubject` being required as well.

Reply:
… ah, https://github.com/openshift/enhancements/pull/1658#discussion_r1735227405 suggests a difference between the `+required` annotation and a `+kubebuilder:validation` annotation, WRT API rules compliance.

I’ll leave K8s experts to review that aspect — I’m primarily interested in whether we want the data, in the first version, to be mandatory or optional _in effect_.

Reply:
Marked it `required` for now, we can change it in future by moving the API to `v1beta2`